src/tools/tool.ts

```diff
@@ -1,22 +1,24 @@
 import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
-import { z, ZodRawShape, ZodTypeAny } from "zod";
+import { z, ZodNever, ZodRawShape } from "zod";
 import { log } from "../logger.js";
 import { CallToolResult } from "@modelcontextprotocol/sdk/types.js";
 import { State } from "../state.js";
 
+export type ToolArgs<Args extends ZodRawShape> = z.objectOutputType<Args, z.ZodNever>;
+
 export abstract class ToolBase<Args extends ZodRawShape> {
     protected abstract name: string;
 
     protected abstract description: string;
 
     protected abstract argsShape: Args;
```

Comment:
```suggestion
export abstract class ToolBase {
    protected abstract name: string;

    protected abstract description: string;

    protected abstract argsShape: ZodRawShape;
    
    protected abstract execute(args: ToolArgs<typeof this.argsShape>): Promise<CallToolResult>;
```
I went on a rabbithole trying to get rid of some redundancy here. Unfortunately I don't think the `Args` generic is doing anything for us.

I would instead get rid of it and use `typeof this.argsShape`. This is still not really going to be a strict type as is. For example in `find.ts` you can define execute to use `ToolArgs<ZodRawShape>` and it'd still pass fine so unfortunately TypeScript isn't that strict/smart anyhow.

Same for `MongoDBToolBase`.

Then we can simply define `argsShape` inside the class as is without replicating it

Reply:
Yeah, I couldn't get typescript to behave exactly how I wanted it to, but figured it's better this way than rewriting the entire server in a proper language 🥲 Anyway, agree with your assessment and have removed the args generic.

Reply:
yeah I was also hoping there's a better way but couldn't come up with an alternative either 🥲